CORE-14575 redpanda: do not destroy scheduling groups

pgellert opened the pull request:
Previously, out custom scheduling groups were created and destroyed on
startup/shutdown. This could cause undefined behaviour if there were
some lingering continuations/timer tasks trying to schedule onto the
destroyed scheduling groups.

While an alternative would be to have a global gate that waits for all
work on each scheduling group to finish, this is infeasible in practice,
because continuations are scheduled from all over the codebase using the
implicit current_scheduling_group(), so it is hard to track down all
usages of a scheduling group, and change all places, including seastar
library functions (such as the tls shutdown code) to use a new global
gate.

Instead, it is simpler and safer to extend the lifetime of the
scheduling groups. We stop destroying them, and let the reactor release
them as it is getting deconstructed.

To avoid multiple instances and metric registration conflicts, we enforce
a singleton pattern with a private constructor. This ensures only one
instance exists per process, used by both production and test code.

Fixes https://redpandadata.atlassian.net/browse/CORE-14575

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert commented on src/v/resource_mgmt/cpu_scheduling.h:
```diff
@@ -22,7 +22,13 @@
+    static ss::future<> create_groups() {
+        if (_initialized) {
+            co_return;
+        }
```
It does not need to be thread safe, because we only initialize once on shard 0 during the application startup sequence.

StephanDollberg commented:
> We make them static here, which ensures that they are only released while the reactor's destructor is called

I guess this works today because ss::scheduling_group is really only a very thing int wrapper which also doesn't have a destructor or anything like that but the ss::scheduling_group outliving the reactor is also kinda a bit odd I find.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--38ff5bc0-6a91-4fdf-8ff3-13e7cdc60951-->
<!--TEST RESULTS-->
<!--dc3d000c-c703-4bdd-a692-9ae18c63336a-->
<!--TEST RESULTS-->
<!--1a0bf097-7624-47f9-a05b-2c7957c03e40-->
<!--TEST RESULTS-->
<!--5d263fdc-0d31-4107-91a8-796d52bdf110-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76113">build#76113</a></summary>


| test_class                          | test_method                     | test_arguments               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                     |
|:------------------------------------|:--------------------------------|:-----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ScalingUpTest                       | test_moves_with_local_retention | {"use_topic_property": true} | integration | https://buildkite.com/redpanda/redpanda/builds/76113#019a7850-7558-4c89-b1ac-6c947f9dbc07 | FLAKY         | 20/21    | upstream reliability is '98.05825242718447'. current run reliability is '95.23809523809523'. drift is 2.82016 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ScalingUpTest&test_method=test_moves_with_local_retention      |
| WriteCachingFailureInjectionE2ETest | test_crash_all                  | {"use_transactions": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/76113#019a7850-754d-47f3-97ea-a7c255704d0c | FLAKY         | 18/21    | upstream reliability is '91.03182256509162'. current run reliability is '85.71428571428571'. drift is 5.31754 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76327">build#76327</a></summary>


| test_class                 | test_method                                  | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:---------------------------|:---------------------------------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| SegmentMsTest              | test_segment_rolling_with_retention_consumer | null               | integration | https://buildkite.com/redpanda/redpanda/builds/76327#019a834e-04a5-487d-8407-88ad8ef341af | FLAKY         | 18/21    | upstream reliability is '92.45283018867924'. current run reliability is '85.71428571428571'. drift is 6.73854 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| ShadowLinkingRandomOpsTest | test_node_operations                         | {"failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/76327#019a834e-04a5-487d-8407-88ad8ef341af | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77023">build#77023</a></summary>


| test_class              | test_method                           | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                         |
|:------------------------|:--------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest | test_simple_remount                   | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/77023#019ac03e-f5e7-4358-bb9d-7fc36e9cb511 | FLAKY         | 16/21    | upstream reliability is '79.01353965183753'. current run reliability is '76.19047619047619'. drift is 2.82306 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount            |
| MultiRestartTest        | test_recovery_after_multiple_restarts | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/77023#019ac03e-f5ea-44bd-96fe-3e5d618bc373 | FLAKY         | 17/21    | upstream reliability is '95.22388059701493'. current run reliability is '80.95238095238095'. drift is 14.2715 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MultiRestartTest&test_method=test_recovery_after_multiple_restarts |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77247">build#77247</a></summary>


| test_class               | test_method                           | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                 |
|:-------------------------|:--------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_mount_inexistent                 | null                              | integration | https://buildkite.com/redpanda/redpanda/builds/77247#019ae445-d538-4461-97ea-04ac0858f75f | FLAKY         | 19/21    | upstream reliability is '99.19571045576407'. current run reliability is '90.47619047619048'. drift is 8.71952 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent                    |
| NodesDecommissioningTest | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true} | integration | https://buildkite.com/redpanda/redpanda/builds/77247#019ae445-d538-4461-97ea-04ac0858f75f | FLAKY         | 18/21    | upstream reliability is '97.05882352941177'. current run reliability is '85.71428571428571'. drift is 11.34454 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
</details>
<!--/body-->

pgellert committed:
redpanda: start sr, pp, audit if scheduling groups

In various fixture tests we pass in a fixed scheduling group to avoid
recreating them or having multiple of them during multi-node fixture
tests.

This injected scheduling groups, however, has also shortcut the
application::initialize function too eagerly, too early, and would avoid
initializing the pandaproxy, schema registry, audit logging.

This commit refactors the code to remove this implicit coupling.

The test behaviour is changed such that the audit logging client is now
always started.

However, the test behaviour is NOT changed in that pandaproxy and schema
registry is not started in these fixture tests, as enabling them
currently causes a range of assertion failures.

pgellert committed:
redpanda: do not destroy scheduling groups

Previously, out custom scheduling groups were created and destroyed on
startup/shutdown. This could cause undefined behaviour if there were
some lingering continuations/timer tasks trying to schedule onto the
destroyed scheduling groups.

While an alternative would be to have a global gate that waits for all
work on each scheduling group to finish, this is infeasible in practice,
because continuations are scheduled from all over the codebase using the
implicit current_scheduling_group(), so it is hard to track down all
usages of a scheduling group, and change all places, including seastar
library functions (such as the tls shutdown code) to use a new global
gate.

Instead, it is simpler and safer to extend the lifetime of the
scheduling groups. We stop destroying them, and let the reactor release
them as it is getting deconstructed.

To avoid multiple instances and metric registration conflicts, we enforce
a singleton pattern with a private constructor. This ensures only one
instance exists per process, used by both production and test code.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/0feb54e856a2040b871d526e4de8c4fef748de48..601474c1f4bc11109a154b177b5f649b03ed75e5): rework the implementation to use a static instance only for testing and clean up some broken logic in `initialize` as part of that

pgellert commented on src/v/resource_mgmt/cpu_scheduling.h:
```diff
@@ -106,42 +112,25 @@ class scheduling_groups final {
+    static ss::scheduling_group admin_sg() { return _admin; }
+    static ss::scheduling_group raft_recv_sg() { return _raft_recv; }
+    static ss::scheduling_group kafka_sg() { return _kafka; }
+    static ss::scheduling_group cluster_sg() { return _cluster; }
```
Makes sense. I've reworked this now such that we only have a static instance for testing, but the `scheduling_groups` object is not static itself.

pgellert commented on src/v/cluster/tests/cluster_test_fixture.h:
```diff
@@ -95,7 +95,6 @@ class cluster_test_fixture {
 
     cluster_test_fixture()
```
Yeah, I had to rework this a bit now to ensure that all tests always reuse a single `scheduling_groups` instance. This worked here because of the initialization guard earlier, but now I replaced that with a static singleton instance instead for testing, as that seems simpler.

pgellert commented:
> > We make them static here, which ensures that they are only released while the reactor's destructor is called
> 
> I guess this works today because ss::scheduling_group is really only a very thing int wrapper which also doesn't have a destructor or anything like that but the ss::scheduling_group outliving the reactor is also kinda a bit odd I find.

I guess it's the usual reference lifetime problem. `ss::scheduling_group` is just a reference/handle to a scheduling group, which becomes invalid once the reactor is destroyed (which never happens for us, because we only ever have 1 reactor created per process). I think the bigger problem is that it's too easy to leak scheduling groups into continuations/timers, which makes scheduling groups hard to destroy, and that forces us to extend their lifetime here. But I don't really have a good (upstream) solution to that.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/0feb54e856a2040b871d526e4de8c4fef748de48..23c8865e433e438999c58d6553095acf463d1f4f): rework the implementation to use a static instance only for testing and clean up some broken logic in initialize as part of that

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/23c8865e433e438999c58d6553095acf463d1f4f..aa3b2aee6d4c235e04518d1435583c0fbb11b207): fix build failures introduced in the refactor

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/aa3b2aee6d4c235e04518d1435583c0fbb11b207..d0a6c83a956b2d762f0b8fdd95316bb91da37784): fix remaining test failures introduced by the refactoring + not destroying scheduling groups

nguyen-andrew commented on src/v/redpanda/tests/fixture.cc:
```diff
@@ -118,10 +118,15 @@ redpanda_thread_fixture::redpanda_thread_fixture(
-      schema_reg_config(schema_reg_port),
-      proxy_client_config(kafka_port),
+      proxy_port.transform([this](auto port) { return proxy_config(port); }),
+      proxy_port.and_then([this, kafka_port](auto) {
```
Why do we use `and_then` + `make_optional` here compared to just `transform` like the previous line? Same with line 127

nguyen-andrew replied:
Is it because the lambda isn't actually using the proxy_port?

pgellert replied:
> Is it because the lambda isn't actually using the proxy_port?

Yeah exactly. I guess transform would work just as well, but I was going for this distinction that the port is not used there.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/d0a6c83a956b2d762f0b8fdd95316bb91da37784..40f09be2e6dab7750b5bab66e42cf6ef2466a989): rebase to dev to fix conflicts

pgellert commented:
@travisdowns @StephanDollberg when you have a chance, can I get a re-review on this please? If it looks good, I'd like to get this into dev.

StephanDollberg reviewed (commented):
Looks fine to me but Travis had comments so should check with him he considers them addressed.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/40f09be2e6dab7750b5bab66e42cf6ef2466a989..607acb1a615972e567699acd7d07a045f0c81a36): refactor to use a single static `scheduling_groups` instance in the production path as well (not just for tests) after discussions with @travisdowns on slack.
